2019.10+fio devel.fpga.pr

ldts opened the pull request:
Please do not submit a Pull Request via github.  Our project makes use of
mailing lists for patch submission and review.  For more details please
see https://www.denx.de/wiki/U-Boot/Patches

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -95,7 +378,6 @@ int board_init(void)
 #ifdef CONFIG_FSL_QSPI
 	board_qspi_init();
 #endif
-
```
No need to delete here as part of this patch.

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +126,250 @@ int board_early_init_f(void)
+/* stub TEE operations until ready to integrate with it*/
+static struct tee_hash t_hash;
+
+static int optee_get_tee_hash(uint8_t *tee_hash, int *tee_hash_len, int *payload_len)
```
Should we conditionally enable the optee check by making sure CONFIG_OPTEE is enabled?

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -13,11 +13,50 @@
+#define M4_SIZE			(SZ_128K + SZ_64K)
+#define M4_ENTRY_OFFSET		0x1004
+#define M4_WATERMARK_OFFSET	0x1000
+#define M4_WATERMARK(fw)						\
```
Looks like most pieces here are 7ulp generic, might be good to move to the common files and just call the right functions here.

ldts replied:
done

ricardosalveti commented:
Can be tested with https://github.com/foundriesio/meta-lmp/pull/92

mike-scott commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +126,250 @@ int board_early_init_f(void)
+	writel(M4_ENTRY(M4_BASE), SIM0_RBASE + 0x70);
+}
+
+int fpga_loadbitstream(int devnum, char *fpgadata, size_t size,
```
Can we add a comment here to clarify why we're using FPGA functions?  Something worded a bit better, but to the extent of: Leverage the FPGA API in u-boot to trigger loading of the M4 firmware to QSPI

mike-scott commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +126,250 @@ int board_early_init_f(void)
+	printf("\tM4 Sram : 0x%x\n", M4_BASE);
+	printf("\tM4 Size : 0x%x\n", M4_SIZE);
+
+	if (optee_get_tee_hash(tee_hash, &tee_hash_len, &tee_payload_len)) {
```
If we move this call up above the `spi_flash_read` call.  We can use `tee_payload_len` as the size to read from QSPI (rounded up of course) instead of hard-coded to `M4_SIZE`.

mike-scott commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +126,250 @@ int board_early_init_f(void)
+	uint8_t tee_hash[FIT_MAX_HASH_LEN];
+	uint8_t hash[FIT_MAX_HASH_LEN];
+
+	if (spi_flash_read(flash, 0, M4_SIZE, (void *) M4_BASE)) {
```
I think this could be changed to a memory clear (memset) statement using M4_SIZE.  (See below about shifting of `optee_get_tee_hash` to here (under the memset).

Then `spi_flash_read` using the `tee_payload_len` value for the size (rounded up).

This would avoid loading the entire SRAM space from QSPI which could be an attack vector as the later hash check only verifies the hash of up to `tee_payload_len`.

mike-scott commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +126,250 @@ int board_early_init_f(void)
+
+	printf("\tM4 Entry: 0x%x\n", M4_ENTRY(M4_BASE));
+	printf("\tM4 Sram : 0x%x\n", M4_BASE);
+	printf("\tM4 Size : 0x%x\n", M4_SIZE);
```
Size using `tee_payload_len`

ldts replied:
fixed

mike-scott reviewed (changes requested):
Couple of thoughts.

mike-scott commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
+	/* do not boot if the FPGA is corrupted (ie, sha different than TEE's) */
+	if (calculate_hash((const void *) M4_BASE, tee_payload_len, "sha256",
+			   hash, &hash_len)) {
+		printf("M4: cant boot, unsupported hash algorithm\n");
```
@ldts Hm.. so at this point we've loaded the binary into memory and if we can't boot for some reason we return here and at 2 other points below.

But another process could find a way to `writel(M4_ENTRY(M4_BASE), SIM0_RBASE + 0x70);` and the M4 would boot the firmware we just said was bad.

Should we establish an error path here which does another `memset` to make sure we don't leave the binary in memory (would also use that path for the 2 returns below).

I might be over thinking this.. @ricardosalveti what do you think?

ricardosalveti replied:
Are we booting from memory or from QSPI here? Because at this point the image is already flashed in QSPI, so not much to do to protect that (unless we decide to erase it, which could work as well, since the next reboot will force a new flash to happen).

mike-scott replied:
@ricardosalveti @ldts so maybe create a standard error path which erases the QSPI on unrecoverable issues like this (and the few below it?) -- is there a danger of just erasing the QSPI over and over?

mike-scott commented:
@ldts Looking better: we just need to discuss the error path for `m4_boot()`.

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
+		printf("M4: Image size too small (%d)!\n", size);
+		return -EINVAL;
+	} else
+		printf("M4: Fit Image Size %d\n", size);
```
No need for the else statement here, just move the printf line after the if block.

ldts replied:
yep

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
+
+	len = round_down(M4_SIZE, mtd->erasesize);
+	if (spi_flash_erase(flash, 0, len))
+		printf("M4: Failed to erase spi flash\n");
```
Should we return -EINVAL when failing to erase the flash?

ldts replied:
yes. I am still working under the paradigm of booting to a shell is best rather than security...need to change my mental mode!

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
 	return 0;
 }
 
+#if defined(CONFIG_OPTEE) && defined(CONFIG_FPGA)
```
Probably a good idea to allow having the update/flash/boot logic still work without OPTEE, as that way we can test it independently. 

You can protect just the TEE stuff (compare hash, etc) with CONFIG_OPTEE, that way we can still test this without OPTEE (which is useful during bringup).

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
+	if (get_boot_mode() == DUAL_BOOT) {
+		/* in DUAL mode, the M4 should be already running */
+		printf("M4: Cant upgrade or boot M4 in dual boot mode\n");
+		return 0;
```
Wonder if we should return an error here, as otherwise it might give the wrong impression that it was able to parse and load the m4 content from fit, which is not the case.

ldts replied:
we cant return an error here since at this point we don't know yet if the user wanted to upgrade. but we can add this additional check to where the upgrade is supposed to initiate and then raise the error there?

mike-scott replied:
This is tricky.  In theory we need to have the A7 negotiate with the M4 to reboot prior to A7 reset (which should put the M4 back into single boot mode).

If we hit this, it's probably one of the following:
- a WDOG event / unexpected reboot in Linux by the A7
- we're boot looping in u-boot after booting the M4 firmware but prior to the boot count limit being reached where we can rollback?

@ricardosalveti I'm not sure there's any value in knowing whether the FIT M4 content is valid at this point.  We can't do anything with it.

ricardosalveti replied:
This function will get called when parsing the fit image (and if the fit includes the fpga entry), so we should cover all the cases you suggested here.

First is what to do if boot mode is not single boot? I would suggest aborting as the user is doing something wrong anyway (loading a fit that includes fpga but booting in a scenario in which u-boot is unable to update m4).

Then we need to make sure the M4 is in a proper state, to cover the A7 reboot scenario. We can always have linux to message M4 during a normal shutdown process, but we will have to cover the unexpected reboot one as you mentioned, as at that point M4 is going to be running while we try to boot/update it. The same logic will also apply to rollback indeed, as we can have failures at both M4 and A7.

ricardosalveti commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
+	}
+
+	/* get the M4 node in the FIT image */
+	noffset = m4_get_fit_node((void *) env_get_ulong("initrd_addr", 16, 0UL));
```
If I read this correctly, this function will only be called after u-boot parses the fit image, in which case we know the fit itself is valid. The issue is that the hash is not provided by it, and we need to use when comparing with optee, but hardcoding the load address to 'initrd_addr' is not ideal (as this can change depending on the boot script).

Not sure how to best handle this here, as it might require core u-boot changes. Maybe we just recalculate the hash from the data and len given via this function call?

ldts replied:
sure, that should lead to the same results and make the code much leaner. or we can extend the api so the hash for submodles are notified to listeners when parsed. I think recalculating the hash is much simpler and maintainable though so I lean towards that.

ldts replied:
@mike-scott what do you think?

mike-scott replied:
@ldts @ricardosalveti I'd lean towards recalculating the hash.

ldts committed:
[FIO extras] avb: remove fastboot dependency

Work around build problems on imx7ulpea

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO extras] avb: avoid imx name collision

ldts committed:
[FIO toup] mmc: fsl_esdhc_imx: initialize data for imx7ulp

changes done as per the linux kernel

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO toup] optee: fix returning CID to TEE

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO toup] rpmb: use cache aligned buffers on route commands

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO toup] imx: mx7ulpea: M4 firmware upgrade

Enables doing an M4 upgrade via the FPGA FIT interface. The only
condition to trigger an upgrade requires that the FIT hash associated
to the new firmware is different to the one already installed in QPSI
flash.

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO extras] mx7ulpea-ucom_lmp_defconfig: enable TEE and AVB

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO toup] imx: mx7ulpea: secure M4 firmware upgrade

Uses a Trusted Application to store the hash associated to the
currently installed firmware. If the hash associated to the firmware
that we wish to program is different to the one stored in the TA, an
upgrade is performed.

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>
